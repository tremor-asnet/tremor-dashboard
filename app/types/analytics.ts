export interface IAnalyticsInfo {
  id: string;
  photo: string;
  name: string;
  description: string;
  price: number;
  location: string;
}

export enum CHART_TYPE {
  SALE = "sale",
  PERFORMANCE = "performance",
}

export type LINE_CHART_DATA = {
  date: string;
  "Desktop apps"?: number;
  "Mobile apps"?: number;
};

export type LINE_CHART = {
  id: string;
  display: string;
  descValue: string;
  desc: string;
  modified: string;
  data: LINE_CHART_DATA[];
};

type TFixedProps = {
  eventType: "dot" | "category" | "bar" | "slice" | "bubble";
  categoryClicked: string;
};

type TBaseEventProps = TFixedProps & {
  [key: string]: number | string;
};

export type TEventProps = TBaseEventProps | null | undefined;
