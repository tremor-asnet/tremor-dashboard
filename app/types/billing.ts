export interface Invoice {
  date: string;
  code: string;
  price: number;
}

export interface Transaction {
  id: number;
  createdAt: string;
  service: string;
  amount: number;
  type: number;
  status: number;
}
export interface BillingCardData {
  id: number;
  type: string;
  expire: string;
  cardNumber: string;
  holderFullName: string;
  cardLast4Digit: string;
}
