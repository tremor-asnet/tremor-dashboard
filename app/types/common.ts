export interface Step {
  index: number;
  content: string;
}

export interface IMedia {
  image: string;
}

export interface CdnResponse {
  data: {
    width: number;
    height: number;
    image: ImageInfo;
  };
}

export interface ImageInfo {
  filename: string;
  url: string;
}

export interface SortItem {
  key: string;
  direction: string;
}

export type TSearchParams = {
  query?: string;
  isAvailable?: string;
  status?: string;
  page?: number;
  sortBy?: string;
  orderBy?: string;
};
