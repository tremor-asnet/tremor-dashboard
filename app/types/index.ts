export * from "./user";
export * from "./breadcrumb";
export * from "./tab";
export * from "./project";
export * from "./analytics";
export * from "./sales";
export * from "./profile";
export * from "./table-list";
export * from "./orders";
