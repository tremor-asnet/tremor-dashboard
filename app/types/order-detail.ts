export type ProductDetail = {
  id: number;
  name: string;
  count: number;
  price: number;
  url: string;
};

export type TOrderDetail = {
  id: number;
  order_code: string;
  order_deliver_price: number;
  order_tax: string;
  created_at: string;
  products: ProductDetail[];
  track_order_info: {
    generate_order_id: string;
    generate_order_at: string;
    transmited_to_courier_at: string;
    delivered_at: string;
  };
  billing_info: {
    last4_bank: string;
    owner_name: string;
    company_name: string;
    email: string;
    vat: string;
  };
};
