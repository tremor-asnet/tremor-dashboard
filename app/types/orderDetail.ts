import { ReactNode } from "react";

export interface StepOrderProps {
  iconInfo: ReactNode;
  titleInfo: string;
  descriptionInfo?: string;
  className?: string;
  backgroundColor?: string;
}
