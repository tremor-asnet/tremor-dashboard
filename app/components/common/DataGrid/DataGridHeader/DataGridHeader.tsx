// Components
import { HeaderCellContents } from "@/components";
import { TableHead, TableHeaderCell, TableRow } from "@tremor/react";

// Types
import { ColumnType } from "@/types";
import { SortItem } from "@/components/Table/TableProduct/TableProduct";

interface DataTableHeaderProps<T> {
  columns: ColumnType<T>[];
  onHeaderClick?: (keyColumn: string) => void;
  sortKey?: string;
  sortDirection?: string;
}

const DataGridHeader = <T,>({
  columns,
  onHeaderClick,
  sortKey,
  sortDirection,
}: DataTableHeaderProps<T>) => {
  return (
    <TableHead>
      <TableRow>
        {columns.map(({ key, title, sortable }) => {
          const handleHeaderClick = () => {
            sortable ? onHeaderClick?.(key) : null;
          };

          return (
            <TableHeaderCell
              onClick={handleHeaderClick}
              key={key}
              className="px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer">
              <HeaderCellContents
                title={title}
                keyColumn={key}
                sortKey={sortKey}
                sortDirection={sortDirection}
                sortable={sortable}
              />
            </TableHeaderCell>
          );
        })}
      </TableRow>
    </TableHead>
  );
};

export default DataGridHeader;
