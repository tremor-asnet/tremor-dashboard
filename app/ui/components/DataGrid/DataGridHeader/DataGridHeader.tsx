"use client";

import { useCallback, useState } from "react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";

// Components
import { HeaderCellContents } from "@/ui/components";
import { TableHead, TableHeaderCell, TableRow } from "@tremor/react";

// Types
import { ColumnType } from "@/types";

// Constants
import { DIRECTION } from "@/constants";

interface DataTableHeaderProps<T> {
  columns: ColumnType<T>[];
}

const DataGridHeader = <T,>({ columns }: DataTableHeaderProps<T>) => {
  const [sortField, setSortField] = useState<string>("");
  const [sortType, setSortType] = useState<string>("");
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();
  const params = new URLSearchParams(searchParams);

  const handleSortingChange = useCallback(
    (key: string, urlParams: URLSearchParams) => {
      return () => {
        setSortField(key);
        urlParams.set("sortBy", key);

        const orderByParam =
          urlParams.get("orderBy") === DIRECTION.DESC
            ? DIRECTION.ASC
            : DIRECTION.DESC;

        urlParams.set("orderBy", orderByParam);
        setSortType(orderByParam);

        // Update url param
        replace(`${pathname}?${urlParams.toString()}`);
      };
    },
    [],
  );

  return (
    <TableHead>
      <TableRow>
        {columns.map(({ key, title, isSortable = false }) => {
          return (
            <TableHeaderCell
              onClick={
                isSortable ? handleSortingChange(key, params) : undefined
              }
              key={key}
              className="px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer print:!text-primary dark:print:!text-primary dark:print:opacity-100">
              <HeaderCellContents
                title={title}
                keyColumn={key}
                sortField={sortField}
                sortType={sortType}
                isSortable={isSortable}
              />
            </TableHeaderCell>
          );
        })}
      </TableRow>
    </TableHead>
  );
};

export default DataGridHeader;
