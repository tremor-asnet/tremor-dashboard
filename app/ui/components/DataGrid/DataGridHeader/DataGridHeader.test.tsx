import { RenderResult, fireEvent, render } from "@testing-library/react";

// Types
import { ColumnType, Product } from "@/types";

// Components
import DataGridHeader from "./DataGridHeader";

// Mock next
jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(),
  usePathname: jest.fn(),
  useRouter: () => ({ replace: jest.fn() }),
}));

const mockColumns: ColumnType<Product>[] = [
  {
    key: "id",
    title: "Id",
    isSortable: true,
  },
  {
    key: "productName",
    title: "Product",
  },
];

describe("DataGridHeader component", () => {
  let renderedComponent: RenderResult;

  beforeEach(() => {
    renderedComponent = render(
      <table>
        <DataGridHeader columns={mockColumns} />
      </table>,
    );
  });

  it("Should matches snapshot", () => {
    const { container } = renderedComponent;
    // Take a snapshot of the rendered component
    expect(container).toMatchSnapshot();
  });

  it("Should be call handleSorting function when isSortable column header is clicked", () => {
    const { container, getByText } = renderedComponent;
    // Click on a isSortable column header
    fireEvent.click(getByText(mockColumns[0].title));

    // Click on a isSortable column header again
    fireEvent.click(getByText(mockColumns[0].title));

    // Click on a isSortable column header again
    fireEvent.click(getByText(mockColumns[0].title));

    // Click on a non-isSortable column header
    fireEvent.click(getByText(mockColumns[1].title));

    expect(container).toMatchSnapshot();
  });
});
