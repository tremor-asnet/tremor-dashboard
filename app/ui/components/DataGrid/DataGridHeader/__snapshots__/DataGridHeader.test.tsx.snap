// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGridHeader component Should be call handleSorting function when isSortable column header is clicked 1`] = `
<div>
  <table>
    <thead
      class="tremor-TableHead-root text-left text-tremor-content dark:text-dark-tremor-content"
    >
      <tr
        class="tremor-TableRow-row"
      >
        <th
          class="tremor-TableHeaderCell-root whitespace-nowrap text-left top-0 text-tremor-content-strong px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer print:!text-primary dark:print:!text-primary dark:print:opacity-100"
        >
          <div
            class="tremor-Flex-root flex w-full flex-row justify-between items-center"
          >
            <div
              class="tremor-Flex-root flex w-full flex-row justify-between items-center"
            >
              Id
            </div>
            <div
              class="tremor-Flex-root flex w-full flex-col justify-between items-center relative ml-4 -mt-6"
            >
              <div
                class="tremor-Flex-root flex w-full flex-row justify-end items-center absolute top-0"
              >
                <button
                  class="tremor-Button-root shrink-0 inline-flex group font-medium outline-none text-sm text-tremor-brand dark:text-dark-tremor-brand bg-transparent hover:text-tremor-brand-emphasis dark:hover:text-dark-tremor-brand-emphasis justify-center items-center opacity-50 dark:opacity-100"
                >
                  <span
                    class="tremor-Button-text text-tremor-default whitespace-nowrap"
                  >
                    <svg
                      class="w-4 h-4 text-primary"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M7 14l5-5 5 5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
              <div
                class="tremor-Flex-root flex w-full flex-row justify-end items-center absolute top-1.5"
              >
                <button
                  class="tremor-Button-root shrink-0 inline-flex group font-medium outline-none text-sm text-tremor-brand dark:text-dark-tremor-brand bg-transparent hover:text-tremor-brand-emphasis dark:hover:text-dark-tremor-brand-emphasis justify-center items-center opacity-50 dark:opacity-100"
                >
                  <span
                    class="tremor-Button-text text-tremor-default whitespace-nowrap"
                  >
                    <svg
                      class="w-4 h-4 fill-secondary"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </th>
        <th
          class="tremor-TableHeaderCell-root whitespace-nowrap text-left top-0 text-tremor-content-strong px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer print:!text-primary dark:print:!text-primary dark:print:opacity-100"
        >
          <div
            class="tremor-Flex-root flex w-full flex-row justify-between items-center"
          >
            <div
              class="tremor-Flex-root flex w-full flex-row justify-between items-center"
            >
              Product
            </div>
          </div>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`DataGridHeader component Should matches snapshot 1`] = `
<div>
  <table>
    <thead
      class="tremor-TableHead-root text-left text-tremor-content dark:text-dark-tremor-content"
    >
      <tr
        class="tremor-TableRow-row"
      >
        <th
          class="tremor-TableHeaderCell-root whitespace-nowrap text-left top-0 text-tremor-content-strong px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer print:!text-primary dark:print:!text-primary dark:print:opacity-100"
        >
          <div
            class="tremor-Flex-root flex w-full flex-row justify-between items-center"
          >
            <div
              class="tremor-Flex-root flex w-full flex-row justify-between items-center"
            >
              Id
            </div>
            <div
              class="tremor-Flex-root flex w-full flex-col justify-between items-center relative ml-4 -mt-6"
            >
              <div
                class="tremor-Flex-root flex w-full flex-row justify-end items-center absolute top-0"
              >
                <button
                  class="tremor-Button-root shrink-0 inline-flex group font-medium outline-none text-sm text-tremor-brand dark:text-dark-tremor-brand bg-transparent hover:text-tremor-brand-emphasis dark:hover:text-dark-tremor-brand-emphasis justify-center items-center opacity-50 dark:opacity-100"
                >
                  <span
                    class="tremor-Button-text text-tremor-default whitespace-nowrap"
                  >
                    <svg
                      class="w-4 h-4 fill-secondary"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M7 14l5-5 5 5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
              <div
                class="tremor-Flex-root flex w-full flex-row justify-end items-center absolute top-1.5"
              >
                <button
                  class="tremor-Button-root shrink-0 inline-flex group font-medium outline-none text-sm text-tremor-brand dark:text-dark-tremor-brand bg-transparent hover:text-tremor-brand-emphasis dark:hover:text-dark-tremor-brand-emphasis justify-center items-center opacity-50 dark:opacity-100"
                >
                  <span
                    class="tremor-Button-text text-tremor-default whitespace-nowrap"
                  >
                    <svg
                      class="w-4 h-4 fill-secondary"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </th>
        <th
          class="tremor-TableHeaderCell-root whitespace-nowrap text-left top-0 text-tremor-content-strong px-6 py-2 text-[10.4px] leading-[17px] dark:text-white tracking-[0.2px] font-bold opacity-70 uppercase cursor-pointer print:!text-primary dark:print:!text-primary dark:print:opacity-100"
        >
          <div
            class="tremor-Flex-root flex w-full flex-row justify-between items-center"
          >
            <div
              class="tremor-Flex-root flex w-full flex-row justify-between items-center"
            >
              Product
            </div>
          </div>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
